reference/test/solver/batch_cg_kernels.cpp

```diff
@@ -0,0 +1,201 @@
+    using LogData = gko::batch::log::detail::log_data<real_type>;
+    using LinSys = gko::test::LinearSystem<Mtx>;
+
+    BatchCg()
+        : exec(gko::ReferenceExecutor::create()),
+          mat(gko::share(
+              gko::test::generate_3pt_stencil_batch_matrix<const Mtx>(
+                  exec, num_batch_items, num_rows))),
+          linear_system(gko::test::generate_batch_linear_system(mat, num_rhs))
+    {
+        auto executor = this->exec;
+        solve_lambda = [executor](const Settings opts,
+                                  const gko::batch::BatchLinOp* prec,
+                                  const Mtx* mtx, const MVec* b, MVec* x,
+                                  LogData& log_data) {
+            gko::kernels::reference::batch_cg::apply<typename Mtx::value_type>(
+                executor, opts, mtx, prec, b, x, log_data);
+        };
+    }
+
+    std::shared_ptr<const gko::ReferenceExecutor> exec;
+    const real_type eps = 1e-3;
+    const gko::size_type num_batch_items = 2;
+    const int num_rows = 15;
+    const int num_rhs = 1;
+    const Settings solver_settings{100, eps,
+                                   gko::batch::stop::tolerance_type::relative};
+    std::shared_ptr<const Mtx> mat;
+    LinSys linear_system;
+    std::function<void(const Settings, const gko::batch::BatchLinOp*,
+                       const Mtx*, const MVec*, MVec*, LogData&)>
+        solve_lambda;
+};
+
+TYPED_TEST_SUITE(BatchCg, gko::test::RealValueTypes, TypenameNameGenerator);
+
+
+TYPED_TEST(BatchCg, SolvesStencilSystem)
+{
+    auto res = gko::test::solve_linear_system(this->exec, this->solve_lambda,
+                                              this->solver_settings,
+                                              this->linear_system);
+
+    for (size_t i = 0; i < this->num_batch_items; i++) {
+        ASSERT_LE(res.host_res_norm->get_const_values()[i] /
+                      this->linear_system.host_rhs_norm->get_const_values()[i],
+                  this->solver_settings.residual_tol);
+    }
+    GKO_ASSERT_BATCH_MTX_NEAR(res.x, this->linear_system.exact_sol,
+                              this->eps * 10);
+}
+
+
+TYPED_TEST(BatchCg, StencilSystemLoggerLogsResidual)
+{
+    using value_type = typename TestFixture::value_type;
+    using real_type = gko::remove_complex<value_type>;
+
+    auto res = gko::test::solve_linear_system(this->exec, this->solve_lambda,
+                                              this->solver_settings,
+                                              this->linear_system);
+
+    auto res_log_array = res.log_data->res_norms.get_const_data();
+    for (size_t i = 0; i < this->num_batch_items; i++) {
+        ASSERT_LE(
+            res_log_array[i] / this->linear_system.host_rhs_norm->at(i, 0, 0),
+            this->solver_settings.residual_tol);
+        ASSERT_NEAR(res_log_array[i], res.host_res_norm->get_const_values()[i],
+                    10 * this->eps);
```

Comment:
does the residual still use eps not the r<T>?

Reply:
I want to set it to a lower tol so check against that tolerance and not r

Reply:
I am not sure checking lower tolerance on the solution is a good idea because it will depends on the matrix condition number and it is not the criterion can help here.

Reply:
I this test I mainly want to check that the logger logs the residual, so I think it should be fine here.